sklearn/cluster/optics_.py

```diff
@@ -0,0 +1,754 @@
+    "OPTICS: ordering points to identify the clustering structure." ACM SIGMOD
+    Record 28, no. 2 (1999): 49-60.
+    """
+
+    def __init__(self, min_samples=5, max_bound=np.inf, metric='euclidean',
+                 p=2, metric_params=None, maxima_ratio=.75,
+                 rejection_ratio=.7, similarity_threshold=0.4,
+                 significant_min=.003, min_cluster_size_ratio=.005,
+                 min_maxima_ratio=0.001, algorithm='ball_tree',
+                 leaf_size=30, n_jobs=1):
+
+        self.max_bound = max_bound
+        self.min_samples = min_samples
+        self.maxima_ratio = maxima_ratio
+        self.rejection_ratio = rejection_ratio
+        self.similarity_threshold = similarity_threshold
+        self.significant_min = significant_min
+        self.min_cluster_size_ratio = min_cluster_size_ratio
+        self.min_maxima_ratio = min_maxima_ratio
+        self.algorithm = algorithm
+        self.metric = metric
+        self.metric_params = metric_params
+        self.p = p
+        self.leaf_size = leaf_size
+        self.n_jobs = n_jobs
+
+    def fit(self, X, y=None):
+        """Perform OPTICS clustering
+
+        Extracts an ordered list of points and reachability distances, and
+        performs initial clustering using `max_bound` distance specified at
+        OPTICS object instantiation.
+
+        Parameters
+        ----------
+        X : array, shape (n_samples, n_features)
+            The data.
+
+        y : ignored
+
+        Returns
+        -------
+        self : instance of OPTICS
+            The instance.
+        """
+        X = check_array(X, dtype=np.float)
+
+        n_samples = len(X)
+        # Start all points as 'unprocessed' ##
+        self._processed = np.zeros((n_samples, 1), dtype=bool)
+        self.reachability_ = np.ones(n_samples) * np.inf
+        self.core_distances_ = np.ones(n_samples) * np.nan
+        # Start all points as noise ##
+        self.labels_ = -np.ones(n_samples, dtype=int)
+        self.ordering_ = []
+
+        # Check for valid n_samples relative to min_samples
+        if self.min_samples > n_samples:
+            raise ValueError("Number of training samples (n_samples=%d) must "
+                             "be greater than min_samples (min_samples=%d) "
+                             "used for clustering." %
+                             (n_samples, self.min_samples))
+
+        nbrs = NearestNeighbors(n_neighbors=self.min_samples,
+                                algorithm=self.algorithm,
+                                leaf_size=self.leaf_size, metric=self.metric,
+                                metric_params=self.metric_params, p=self.p,
+                                n_jobs=self.n_jobs)
+
+        nbrs.fit(X)
+        self.core_distances_[:] = nbrs.kneighbors(X,
+                                                  self.min_samples)[0][:, -1]
+        self._nbrs = nbrs
```

Comment:
I do not think that it is necessary to store this object. It is used only in the fit. Storing it adds weight to the memory and to the pickle.